src/orc2/mod.rs

```diff
@@ -0,0 +1,2186 @@
+    object_layer: LLVMOrcObjectLayerRef,
+}
+
+#[llvm_versions(12.0..=latest)]
+impl RTDyldObjectLinkingLayer {
+    pub fn get<'a>(&'a self) -> ObjectLayer<'a> {
+        unsafe { ObjectLayer::new_borrowed(self.object_layer) }
+    }
+}
+
+#[llvm_versions(13.0..=latest)]
+#[derive(Debug, Clone)]
+#[repr(transparent)]
+pub struct IRTransformLayer<'jit> {
+    ir_transform_layer: LLVMOrcIRTransformLayerRef,
+    _marker: PhantomData<&'jit ()>,
+}
+
+#[llvm_versions(13.0..=latest)]
+impl<'jit> IRTransformLayer<'jit> {
+    unsafe fn new_borrowed(ir_transform_layer: LLVMOrcIRTransformLayerRef) -> Self {
+        assert!(!ir_transform_layer.is_null());
+        IRTransformLayer {
+            ir_transform_layer,
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn emit<'ctx>(
+        &self,
+        materialization_responsibility: MaterializationResponsibility,
+        module: ThreadSafeModule<'ctx>,
+    ) {
+        unsafe {
+            LLVMOrcIRTransformLayerEmit(
+                self.ir_transform_layer,
+                materialization_responsibility.materialization_responsibility,
+                module.thread_safe_module,
+            );
+        }
+        forget(module);
+    }
+
+    pub fn set_transform() {
+        todo!();
+    }
+}
+
+#[llvm_versions(12.0..=latest)]
+#[llvm_versioned_item]
+#[derive(Debug)]
+pub struct MaterializationUnit<'jit> {
+    materialization_unit: LLVMOrcMaterializationUnitRef,
+    _marker: PhantomData<&'jit ()>,
+}
+
+#[llvm_versions(12.0..=latest)]
+impl<'jit> MaterializationUnit<'jit> {
+    unsafe fn new(materialization_unit: LLVMOrcMaterializationUnitRef) -> Self {
+        assert!(!materialization_unit.is_null());
+        MaterializationUnit {
+            materialization_unit,
+            _marker: PhantomData,
+        }
+    }
+
+    #[llvm_versions(13.0..=latest)]
+    pub fn create<M>(
+        name: &str,
+        mut symbols: SymbolFlagsMapPairs,
+        static_initializer: Option<SymbolStringPoolEntry>,
+        materializer: Box<M>,
+    ) -> Self
+    where
+        M: Materializer + 'jit,
+    {
+        unsafe {
+            MaterializationUnit::new(LLVMOrcCreateCustomMaterializationUnit(
+                to_c_str(name).as_ptr(),
+                Box::into_raw(materializer) as *mut c_void,
+                symbols.raw_ptr(),
+                symbols.len(),
+                static_initializer
+                    .map(|s| s.entry)
+                    .unwrap_or(ptr::null_mut()),
+                <M as UnsafeMaterializer>::materialization_unit_materialize,
+                <M as UnsafeMaterializer>::materialization_unit_discard,
+                <M as UnsafeMaterializer>::materialization_unit_destroy,
+            ))
+        }
+    }
+
+    pub fn from_absolute_symbols(mut symbols: SymbolMapPairs) -> Self {
+        unsafe {
+            MaterializationUnit::new(LLVMOrcAbsoluteSymbols(symbols.raw_ptr(), symbols.len()))
+        }
+    }
```

Comment:
```suggestion
    pub fn from_absolute_symbols(mut symbols: SymbolMapPairs) -> Self {
        for name in symbols.names_iter() {
            unsafe { LLVMOrcRetainSymbolStringPoolEntry(name.entry) };
        }
        unsafe {
            MaterializationUnit::new(LLVMOrcAbsoluteSymbols(symbols.raw_ptr(), symbols.len()))
        }
    }
```

From the LLVM docs for [`LLVMOrcAbsoluteSymbols`](https://llvm.org/doxygen/group__LLVMCExecutionEngineORC.html#ga2fa4d5572b2c93896a2999f5b5e3ca36):

> The Name fields of the array elements are taken to have been retained for this function.